test_conformance/subgroups/subhelpers.h

```diff
@@ -56,219 +66,943 @@ class subgroupsAPI {
 {
-    static bool val(cl_device_id device) { return int64_ok(device); }
-};
-template <> struct TypeCheck<cl_long>
+    for (auto i = 0; i < 3; ++i)
+    {
+        lhs.data.s[i] = rhs;
+    }
+}
+
+// set scalar value to scalar
+template <typename Ty>
+typename std::enable_if<std::is_scalar<Ty>::value>::type
+set_value(Ty &lhs, const cl_ulong &rhs)
 {
-    static bool val(cl_device_id device) { return int64_ok(device); }
-};
-template <> struct TypeCheck<cl_float>
+    lhs = static_cast<Ty>(rhs);
+}
+
+// set scalar value to half scalar
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value>::type
+set_value(Ty &lhs, const cl_ulong &rhs)
 {
-    static bool val(cl_device_id) { return true; }
-};
-template <> struct TypeCheck<cl_half>
+    lhs.data = rhs;
+}
+
+
+// compare for common vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
     {
-        return is_extension_available(device, "cl_khr_fp16");
+        if (lhs.s[i] != rhs.s[i])
+        {
+            return false;
+        }
     }
-};
-template <> struct TypeCheck<double>
+    return true;
+}
+
+
+// compare for vectors 3
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_size3::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
 {
-    static bool val(cl_device_id device)
+    for (auto i = 0; i < 3; ++i)
     {
-        int error;
-        cl_device_fp_config c;
-        error = clGetDeviceInfo(device, CL_DEVICE_DOUBLE_FP_CONFIG, sizeof(c),
-                                (void *)&c, NULL);
-        if (error)
+        if (lhs.data.s[i] != rhs.data.s[i])
         {
-            log_info(
-                "clGetDeviceInfo failed with CL_DEVICE_DOUBLE_FP_CONFIG\n");
             return false;
         }
-        return c != 0;
     }
-};
+    return true;
+}
+
+// compare for half vectors
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_vector_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    const int size = sizeof(Ty) / sizeof(typename TypeManager<Ty>::scalar_type);
+    for (auto i = 0; i < size; ++i)
+    {
+        if (lhs.data.s[i] != rhs.data.s[i])
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+// compare for scalars
+template <typename Ty>
+typename std::enable_if<std::is_scalar<Ty>::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs == rhs;
+}
+
+// compare for scalar halfs
+template <typename Ty>
+typename std::enable_if<TypeManager<Ty>::is_sb_scalar_type::value, bool>::type
+compare(const Ty &lhs, const Ty &rhs)
+{
+    return lhs.data == rhs.data;
+}
+
+
+inline bool isnan_half(const subgroups::cl_half &x)
```

Comment:
This and a few others are in imageHelpers.h: https://github.com/KhronosGroup/OpenCL-CTS/blob/24e6a9125c44ef9fe42603b605a685af9acb4bdf/test_common/harness/imageHelpers.h#L712
Might consider moving those to the harness as well.

Reply:
fixed in https://github.com/KhronosGroup/OpenCL-CTS/pull/1074